fvm/state/transaction_state.go

```diff
@@ -198,28 +223,48 @@ func (s *TransactionState) CommitParseRestricted(
 		)
 	}
 
-	err := s.mergeIntoParent()
-	if err != nil {
-		return nil, err
+	return s.mergeIntoParent()
+}
+
+// Pause detaches the current nested transaction from the parent transaction,
+// and returns the paused nested transaction state.  The paused nested
+// transaction may be resume via Resume (or AttachAndCommit).
+//
+// WARNING: Pause and Resume are intended for implementing continuation passing
+// style behavior.  The paused nested transaction should not be reused across
+// transactions.
+func (s *TransactionState) Pause(
+	expectedId NestedTransactionId,
+) (
+	*State,
+	error,
+) {
+	if !s.IsCurrent(expectedId) {
+		return nil, fmt.Errorf(
+			"cannot pause unexpected nested transaction: id mismatch",
+		)
+	}
+
+	if s.IsParseRestricted() {
+		return nil, fmt.Errorf(
+			"cannot Pause parse restricted nested transaction")
 	}
 
-	return currentFrame.state, nil
+	return s.pop("pause")
 }
 
-// AttachAndCommitParseRestricted commits the changes in the cached nested
-// transaction to the current (nested) transaction.
-func (s *TransactionState) AttachAndCommitParseRestricted(
-	cachedNestedTransaction *State,
-) error {
-	s.nestedTransactions = append(
-		s.nestedTransactions,
-		nestedTransactionStackFrame{
-			state:            cachedNestedTransaction,
-			parseRestriction: nil,
-		},
-	)
+// Resume attaches the paused nested transaction (state) to the current
+// transaction.
+func (s *TransactionState) Resume(pausedState *State) {
+	s.push(pausedState, nil)
```

Comment:
can `pausedState` be committed already? if so, we may want to manually mark `committed` as false so that it can be accessed.

Reply:
right now, we can resume a committed state (but the only action you can perform in that case is commit).  i'm thinking of adding a wrapper around *State to prevent this case.

Reply:
just fyi.  I'll add the PausedNestedTransaction wrapper later today / re-request another review when it's ready.

Reply:
as discussed on vc, i'll leaves this as is for now and will add the wrapper in a follow up pr.